Unlock encrypted root filesystem via SSH

andrey42 opened the pull request:
This commit add new feature for Debian-based distributions to
 unlock encrypted root partition over SSH This feature is very handy on any
 headless NAS or VPS cloud servers. To use this feature, you will need to
 install dropbear-initramfs package during the system setup.

Signed-off-by: Andrey Prokopenko <job@terem.fr>

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
Many NAS and cloud-based servers requiring encrypted root filesystem do not have displays or keyboard connected, hence the usage of encrypted partitions is limited to servers with physical access to keyboard or the usage of rather ugly and insecure schemas where key is saved along with the initramfs image, making it easily prone for breaking into. Proposed pull request removes this vulnerability for Debian based systems and makes usage of encrypted root filesystems on headless systems, requiring encryption, much more convenient.

### Description
New section inside initramfs zfs setup script checks for the presense of working dropbear SSH process ( to be installed via dropbear-initramfs package) and if it's present, opens two in and out named pipes, then waits for the password input and sends the pool decryption result back.
On the login via SSH, if dropbear module is present and running, the user is presented with password prompt. Once the password is accepted, dropbear initramfs process will be killed and user is requested to reconnect via normal user credentials.

### How Has This Been Tested?
The standard set of zfs-linux debian packages was built via Github Actions free tier build based on my workflow, see more details here: https://github.com/andrey42/zfs-debian/blob/master/.github/workflows/debian-packages-build.yml
Two VM with encrypted partitions were created using aforementioned packages to test booting process with and without encrypted root filesystem. Test scripts for hosting provider are available here: https://github.com/andrey42/zfs-hetzner-vm
Two complete scripts for creating barebone installation using deboostrap, experimental version of ZFS packages and dropbear-initramfs with SSH keys config on Debian 10 and Ubuntu 18.04  are located here: https://github.com/andrey42/zfs-hetzner-vm/tree/vmtest

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

ghfields commented on contrib/initramfs/scripts/zfs.in:
```diff
@@ -424,6 +424,34 @@ decrypt_fs()
 				$ZFS load-key "${ENCRYPTIONROOT}"
 
+                        # Wait for input from named pipe, if Dropbear is running
+                        elif cat /run/dropbear.pid &>/dev/null; then
```
Is it safe to assume if dropbear is ever running that it must be for ZFS root unlock?  If not, would setting a user property on `$ENCRYPTIONROOT` to signal intention be helpful?

andrey42 replied:
Good question.
We do can detect the presence of the attached GPU, because during the boot first active GPU assigned as active tty. For example on some hosters with KVM virtualization machines have these fake boot tty attached, they use this feature as a part of their recovery measures,  but the output goes into the void, unless the owner activated the boot console. Quemu and libvirt follows these rules.
Here is a basic working example on detecting attached GPU:
````
if find /sys/devices/ -name *boot*vga* | wc -l | grep  1 > /dev/null; then
  echo boot tty attached; 
fi
````
I've deliberately excluded this part from the zfs code, because I assume the user installing dropbear-initramfs package makes a conscious choice over accessing the system via SSH, The password request is activated only if zfs loader detects the encrypted partition. In non-standard situation the user has the ability to break the password request, drop into the SSH prompt, do necessary repair/rescue actions, then enter `zfsunlock` command to perform root pool decryption later. In any case, max 3 decryption attempts are accepted, then system forcibly reboots to delay brute password guessing.

felixdoerre commented:
Just as an idea: if the target is debian, why not use the `/lib/cryptsetup/askpass` utility from the `cryptsetup-bin` package? This utility already has the functionality of simultaneously asking for the password on the console, plymouth, systemd and a fifo, which can be used from a dropbear-ssh session with `cryptroot-unlock`. Thereby we always try to ask on the console and never accidently lock-out an user (who may have installed `dropbear-initramfs` but not set up an sshkey for dropbear yet)

andrey42 commented:
> Just as an idea: if the target is debian, why not use the `/lib/cryptsetup/askpass` utility from the `cryptsetup-bin` package? This utility already has the functionality of simultaneously asking for the password on the console, plymouth, systemd and a fifo, which can be used from a dropbear-ssh session with `cryptroot-unlock`. Thereby we always try to ask on the console and never accidently lock-out an user (who may have installed `dropbear-initramfs` but not set up an sshkey for dropbear yet)

Unfortunately this is not a normal login, its an SSH session, hence tty misses graphical features. Also, zfs script already uses systemd-ask-password utility for non-SSH unlock features, hence I'ce decided to reuse it to minimize changes to initramfs hooks.

andrey42 commented:
Tiny update on Dropbear detection mechanism. As per advices given,  some Debian derivatives do not have /var/run folder present inside initramfs, so I've changed dropbear process detection to classic ps | grep approach.

sam-lunt commented:
I feel like this over-complicates the process. You are installing the client script (`zfsunlock`) using an initramfs hook, but that assumes too much about the client (that it's building using initramfs-tools and has zfs installed).

Other setups (ArchLinux's mkinitcpio hook for unlocking zfs remotely, for example) simply wait for the user to ssh in and manually unlock the rootfs. That seems easier, since it allows for any client with the appropriate key to login and unlock the filesystem. Your setup requires the client to install the zfsunlock script, since you use specially created FIFOs in order to send the data.

I think a simpler solution would be to support a method of configuring the initramfs such that it confirms there is an ssh daemon running and then simply waits for a client to login and unlock any drives that it doesn't have keys for.

andrey42 commented:
> I feel like this over-complicates the process. You are installing the client script (`zfsunlock`) using an initramfs hook, but that assumes too much about the client (that it's building using initramfs-tools and has zfs installed).
> 
> Other setups (ArchLinux's mkinitcpio hook for unlocking zfs remotely, for example) simply wait for the user to ssh in and manually unlock the rootfs. That seems easier, since it allows for any client with the appropriate key to login and unlock the filesystem. Your setup requires the client to install the zfsunlock script, since you use specially created FIFOs in order to send the data.
> 
> I think a simpler solution would be to support a method of configuring the initramfs such that it confirms there is an ssh daemon running and then simply waits for a client to login and unlock any drives that it doesn't have keys for.
This approach assumes user is prompted with password login with option to cancel it, it needed.
Provided correct version of dropbear-initramfs package is installed and user has the same set of host keys shared between ssh and initramfs, no special procedures or separate setup (like the one spread across numerous incompetent "guides" offering separate port for initramfs SSH prompt etc.) or other special knowledge or learning from user side is required whatsoever. Manual unlock requires user to know what command to enter, my approach is more "end user friendly". For examples of proper setup see test setup script I've developed for Ubuntu 18 and Debian 10.
For non-encrypted root pool, the password prompt will be silently ignored, even if dropbear-initramfs package is installed. If user wants to execute service operations, he can always cancel the password prompt and run the procedures manually.

andrey42 commented on contrib/initramfs/scripts/zfs.in:
```diff
@@ -424,6 +424,35 @@ decrypt_fs()
+                        # Wait for input from named pipe, if Dropbear is running
+                        elif ps | grep dropbear | grep -v grep &> /dev/null; then
+                          echo "Dropbear is running, waiting for password input from SSH console"
+                          local zfs_in_pipe=/dev/zfspassinpipe
```
makes sense, will update the script

andrey42 replied:
done, dev ref removed

sam-lunt commented:
> > I feel like this over-complicates the process. You are installing the client script (`zfsunlock`) using an initramfs hook, but that assumes too much about the client (that it's building using initramfs-tools and has zfs installed).
> > Other setups (ArchLinux's mkinitcpio hook for unlocking zfs remotely, for example) simply wait for the user to ssh in and manually unlock the rootfs. That seems easier, since it allows for any client with the appropriate key to login and unlock the filesystem. Your setup requires the client to install the zfsunlock script, since you use specially created FIFOs in order to send the data.
> > I think a simpler solution would be to support a method of configuring the initramfs such that it confirms there is an ssh daemon running and then simply waits for a client to login and unlock any drives that it doesn't have keys for.

> This approach assumes user is prompted with password login with option to cancel it, it needed.
> Provided correct version of dropbear-initramfs package is installed and user has the same set of host keys shared between ssh and initramfs, no special procedures or separate setup (like the one spread across numerous incompetent "guides" offering separate port for initramfs SSH prompt etc.) or other special knowledge or learning from user side is required whatsoever. Manual unlock requires user to know what command to enter, my approach is more "end user friendly". For examples of proper setup see test setup script I've developed for Ubuntu 18 and Debian 10.
> For non-encrypted root pool, the password prompt will be silently ignored, even if dropbear-initramfs package is installed. If user wants to execute service operations, he can always cancel the password prompt and run the procedures manually.

You're right, I misunderstood. The `zfsunlock` script is copied to the bin directory and that that command is added to the end of the root user's `~/.profile` file. That still seems a little fragile (e.g. it's making assumptions about the contents of `PATH`, it's assuming that no one else changes `~/.profile`, it's assuming that if someone ssh-es in during boot it's in order to unlock the drives).

It's also not robust to the situation where multiple filesystems need to be unlocked (i.e. when `ZFS_INITRD_ADDITIONAL_DATASETS` contains at least one drive that needs to be unlocked), since it uses the same `/run/zfsunlockcomplete` file no matter what dataset it's unlocking.

I think it's safe to assume that if users can understand how to install zfs on root, configure the initramfs tools, and ssh into a system running from its initramfs, then they can also learn to run the `zfs load-key` command. I think this adds a lot of complexity for a minor gain in convenience. Personally, I would rather it support the same interface that the Arch Linux mkinitcpio hook does (which just waits for the user to ssh in and manually load the key). That's just my opinion as a user though.

andrey42 commented:
> It's also not robust to the situation where multiple filesystems need to be unlocked (i.e. when `ZFS_INITRD_ADDITIONAL_DATASETS` contains at least one drive that needs to be unlocked), since it uses the same `/run/zfsunlockcomplete` file no matter what dataset it's unlocking.

Thanks for the error noted, I will move  /run/zfsunlockcomplete after the cycle for unlocking all the pools has been completed. As for the unlocking multiple pools, so far the script assumes only the root pool requires decryption. I'll extend the script so it will ask in cycle for several keys if there is more than one encrypted pool requiring key present.

andrey42 commented:
New commit version now can recognize and prompt for several encrypted pools password.
Each password prompt will show unique pool name to let user distinguish one pool from another.
If there are no encrypted pools, no prompt will appear. After all pools are unlocked, the boot sequence is resumed, dropbear process and subprocesses will be killed by init-bottom/dropbear script, the user is requested to reconnect to main rootfs ssh daemon.
To break out of password prompt inside initramfs user can simply press Ctrl-C, then later, enter zfsunlock command to repeat the password entry. To delay password brufeforce attack, after 3 consecutive  password errors against any encrypted pool the system will automatically reboot.

andrey42 commented:
> You're right, I misunderstood. The `zfsunlock` script is copied to the bin directory and that that command is added to the end of the root user's `~/.profile` file. That still seems a little fragile (e.g. it's making assumptions about the contents of `PATH`, it's assuming that no one else changes `~/.profile`, it's assuming that if someone ssh-es in during boot it's in order to unlock the drives).

in new version of the commit zfsunlock now will work even if the user cancel the prompt, go to shell and run zfsunlock again, so even if the profile has been accidentally overwritten by some package, the user can run zfsunlock manually.

andrey42 commented on contrib/initramfs/scripts/zfs.in:
```diff
@@ -424,6 +424,35 @@ decrypt_fs()
+                          echo "Dropbear is running, waiting for password input from SSH console"
+                          local zfs_in_pipe=/dev/zfspassinpipe
+                          local zfs_out_pipe=/dev/zfspassoutpipe
+                          if [[ ! -p $zfs_in_pipe ]]; then
```
fixed in last commit

andrey42 commented on contrib/initramfs/hooks/zfsunlock.in:
```diff
@@ -0,0 +1,48 @@
+if [ -r "${DESTDIR}/sbin/dropbear" ]; then
+  cat >"$DESTDIR"/bin/zfsunlock <<'EOF'
+#!/bin/sh
+if cat /run/dropbear.pid &>/dev/null; then
```
fixed by adding sync section to cope with race condition to let dropbear wait for zfs module init, happens in case of multiple vdevs init procedure

rlaager commented on contrib/initramfs/hooks/zfsunlock.in:
```diff
@@ -0,0 +1,60 @@
+
+. /usr/share/initramfs-tools/hook-functions
+
+if [ -r "${DESTDIR}/sbin/dropbear" ]; then
```
What is the point of this check for /sbin/dropbear?

rlaager commented on contrib/initramfs/hooks/zfsunlock.in:
```diff
@@ -0,0 +1,60 @@
+echo
+echo "Unlocking encrypted ZFS pools"
+echo "Enter correct password or press Ctrl-C to exit into shell"
+echo "After 3 incorrect passwords system will reboot"
```
What is the point of rebooting after 3 attempts?

andrey42 replied:
Most simple countermeasure against guessing password attempts.

rlaager replied:
I think a reasonable sleep would be more than sufficient.

rlaager replied:
Doesn't the user have to authenticate to dropbear anyway? dropbear isn't a wide-open-to-the-Internet SSH server, is it?

rlaager commented on contrib/initramfs/hooks/zfsunlock.in:
```diff
@@ -0,0 +1,60 @@
+. /usr/share/initramfs-tools/hook-functions
+
+if [ -r "${DESTDIR}/sbin/dropbear" ]; then
+  cat >"$DESTDIR"/bin/zfsunlock <<'EOF'
```
Is it reasonably possible to install the zfsunlock script directly, rather than having it be created using a big heredoc in this script?

andrey42 replied:
I've used heredoc to prevent the user to enter zfsunlock command and falsely think there are errors in the boot sequence the case user decides no longe user dropbear-initramfs package, and switch to KVM console. Some of hosting providers offer this feature. Initramfs image is rebuilt after dropbear-initramfs uninstall and zfsunlock will be completely gone.

rlaager replied:
I don't see how that's related. If your script simply `cp`ed/`install`ed  a copy of zfsunlock, it would have the exact same effect as writing it out with a heredoc.

rlaager commented on contrib/initramfs/hooks/zfsunlock.in:
```diff
@@ -0,0 +1,60 @@
+echo -e "resuming boot sequence\nyou will be disconnected now, please reconnect in a while ..."
+EOF
+  chmod 0755 "$DESTDIR"/bin/zfsunlock
+  echo "zfsunlock" >> "$DESTDIR"/etc/profile
```
Appending to /etc/profile feels completely wrong and almost surely needs to be redesigned.

andrey42 replied:
Agree, but for the moment I've got no other ideas how to automatically wait for the password input immediately after SSH login. Any useful hints on this regard ?

rlaager replied:
I can think of a few options:
* Do nothing. Require the user to run a command (like `zfsunlock`).
* If dropbear or the initramfs environment support MOTD functionality, output a message describing the command to run.
* Modify `/root/.ssh/authorized_keys` to use the SSH `command=` parameter to force running the unlock command (assuming dropbear supports this).
* Use something in /etc/profile.d (if supported in the initramfs environment) or /etc/profile to print a message. That's basically the same, but seems less "bad".

andrey42 replied:
Tried to use /etc/motd
Unfortunately, for reasons unknown to me MOTD is completely ignored while running initramfs.
Tested on Debian 10.
I"ve modified the script to completely silence any output if there is no encrrypted partition or unlocking already has completed. Would it suffice the requirements ?

rlaager commented on contrib/initramfs/scripts/zfs.in:
```diff
@@ -424,6 +424,34 @@ decrypt_fs()
 				$ZFS load-key "${ENCRYPTIONROOT}"
 
+                        # Wait for input from named pipe, if Dropbear is running
+			elif [ ! -z "$(ps | grep dropbear | grep -v grep 2> /dev/null)" ]; then
```
If you turn one of the characters into a single-character class, you can avoid the need for the second grep, as the pattern will not match itself: `ps | grep [d]ropbear`

rlaager commented on contrib/initramfs/scripts/zfs.in:
```diff
@@ -972,6 +1000,11 @@ mountroot()
 
+	touch /run/zfsunlockcomplete
+	if [ -e /run/zfspassoutpipe ]; then
+	  echo "==mountfscompleted==" > /run/zfspassoutpipe
```
Why does this have == and the other things do not?

rlaager commented:
> > Just as an idea: if the target is debian, why not use the `/lib/cryptsetup/askpass` utility from the `cryptsetup-bin` package? This utility already has the functionality of simultaneously asking for the password on the console, plymouth, systemd and a fifo, which can be used from a dropbear-ssh session with `cryptroot-unlock`. Thereby we always try to ask on the console and never accidently lock-out an user (who may have installed `dropbear-initramfs` but not set up an sshkey for dropbear yet)
> 
> Unfortunately this is not a normal login, its an SSH session, hence tty misses graphical features. 

What does this mean?

> Also, zfs script already uses systemd-ask-password utility for non-SSH unlock features, hence I'ce decided to reuse it to minimize changes to initramfs hooks.

Using systemd-ask-password for the actual prompting isn't necessarily a problem. But it's unclear to me if you've addressed the point about _simultaneously_ prompting on the console and via dropbear. If this _requires_ someone to SSH into the machine, that's bad. If the network is broken, for example, they will have a mess.

andrey42 commented:
> Using systemd-ask-password for the actual prompting isn't necessarily a problem. But it's unclear to me if you've addressed the point about _simultaneously_ prompting on the console and via dropbear. If this _requires_ someone to SSH into the machine, that's bad. If the network is broken, for example, they will have a mess.

You're right, it has to be addressed too. I'll come back with extended version of the script, covering the scenario with simultaneous prompts appearing on boot console and SSH session.

andrey42 commented:
Updated version now simultaneously prompting on the console and via dropbear. Installing dropbear-initramfs now is optional.

felixdoerre commented:
I think, my point didn't come across really well, so I'll try again. The askpass utility (from Debian's `cryptsetup-bin`) will by default ask on the UI (via plymouth, systemd and directly on the console) and a fifo simultaneously. If e.g the plymouth-prompt fails this has no impact on the other prompting mechanisms. The fifo is intended for supplying the password when logged in via ssh/dropbear. This mechanism seems to be exactly what is needed here to unlock a zfs dataset.

> Unfortunately this is not a normal login, its an SSH session, hence tty misses graphical features.

Yes that's exactly the point of this askpass utility. With the `fifo` password method (https://salsa.debian.org/cryptsetup-team/cryptsetup/-/blob/b2cfa897bfcf0f5b91c5e76757b30814aca40fa7/debian/askpass.c#L460), `askpass` will wait for a password supplied by the `cryptroot-unlock` utility (https://salsa.debian.org/cryptsetup-team/cryptsetup/-/blob/b2cfa897bfcf0f5b91c5e76757b30814aca40fa7/debian/initramfs/cryptroot-unlock#L84) which is intended to be used via dropbear.

The whole askpass program seems to be much more exact and clean than writing program name substrings into files and hoping that no other process line from ps will match. Maybe `cryptroot-unlock` will need to be replaced with something zfs-specific but the `askpass`-utility can probably be (re)used without modifications. Generally handling pids manually in shell scripts (e.g. by sorting) can be pretty unstable.

andrey42 commented:
If anyone wants to test SSH unlock using [Debian 10 live image](https://cdimage.debian.org/debian-cd/current-live/amd64/iso-hybrid/debian-live-10.3.0-amd64-gnome.iso) and VirtualBox you can use test install script I've made for this purpose.
 Create VM (2Gb and 20Gb storage will be enough), load live system, then run this script:
````
wget https://raw.githubusercontent.com/andrey42/zfs-hetzner-vm/vmtest/virtualbox-debian10-zfs-setup.sh -qO- | sudo bash -
````

rlaager commented on contrib/initramfs/hooks/zfsunlock.in:
```diff
@@ -0,0 +1,65 @@
+	zfs_fs_name=$(cat /run/zfs_fs_name)
+	zfs_console_askpwd_cmd=$(cat /run/zfs_console_askpwd_cmd)
+	systemd-ask-password "Encrypted ZFS password for ${zfs_fs_name}:" | \
+	/sbin/zfs load-key $zfs_fs_name
```
This line should be indented.

andrey42 replied:
done, indented last line

rlaager commented on contrib/initramfs/scripts/zfs:
```diff
@@ -426,6 +429,7 @@ decrypt_fs()
 
 			# Prompt with systemd, if active 
 			elif [ -e /run/systemd/system ]; then
+                                echo -n "systemd-ask-password" > /run/zfs_console_askpwd_cmd
```
It appears this is using spaces instead of tabs like the rest of the file.

andrey42 replied:
done, replaced spaces with tabs

Rudd-O commented:
Gotta ask:

Have you looked into `tang` and `clevis`?

The combo ("the system" hereonafter) allows you to remotely unlock any machine's LUKS devices, and given how the components are structured, it should be possible to provide a ZFS key unlock to the boot process in much the same way as LUKS unlock does.

See https://www.golinuxcloud.com/network-bound-disk-encryption-tang-clevis/ for more information.  It's pretty straightforward.  The keys are *not* transmitted over the wire upon unlock -- they are reconstructed on the booting machine's side using factors known to the server.

andrey42 commented:
> 
> 
> Gotta ask:
> 
> Have you looked into `tang` and `clevis`?
> 
> The combo ("the system" hereonafter) allows you to remotely unlock any machine's LUKS devices, and given how the components are structured, it should be possible to provide a ZFS key unlock to the boot process in much the same way as LUKS unlock does.
Don't see the nesessity, it makes system more complex without adding useful functionality. SSH authentication using public key is safe channel, shielded from MITM, can be broken only if your client machine or remote password storage cluster has been compromised. Therefore is's perfectly ok to transmit key during the server boot, when server awaits the input during the boot process.

Automation of the password submission can be done very easily via one tiny bash script, where during the boot server first "knocks" to the key server, then server logins via SSH, checks the conditions and transmits the password. 

Works perfectly, and no need for any "keyvault" bullshit offered by any cloud/hosting provider , as they cannot be trusted by definition anyway. In my opinion, anyone using cloud hosted keyvaults on hosting company premised is and storing critical credentials there is bindly trusting hosting company fairness and loyalty, which I wouldn't.

behlendorf commented:
@andrey42 when you get a chance would you mind rebasing this on the master branch and force updating the PR.

andrey42 commented:
rebased onto master branch

rlaager commented on contrib/initramfs/hooks/Makefile.am:
```diff
@@ -1,10 +1,12 @@
 EXTRA_DIST = \
-	$(top_srcdir)/contrib/initramfs/hooks/zfs.in
+	$(top_srcdir)/contrib/initramfs/hooks/zfs.in \
+	$(top_srcdir)/contrib/initramfs/hooks/zfsunlock.in
```
I don't see any substitutions in `zfsunlock.in`. Can't this just be `zfsunlock` in the source tree? Then you shouldn't need it in `EXTRA_DIST`. You will need to change `$(hooks_SCRIPTS):%:%.in` to something like `$(hooks_SCRIPTS):zfs.in`.

rlaager commented on contrib/initramfs/hooks/zfsunlock.in:
```diff
@@ -0,0 +1,65 @@
+. /usr/share/initramfs-tools/hook-functions
+
+if [ -r "${DESTDIR}/sbin/dropbear" ]; then
+	cat >"$DESTDIR"/bin/zfsunlock <<'EOF'
```
Rather than having the `zfsunlock` initramfs hook script write out `/bin/zfsunlock` with `cat` and a HERE doc, could we just install what is the `/bin/zfsunlock` script somewhere on the system (whether that's in `/bin`, or `/usr/share/initramfs-tools/...` is a separate question) and then `cp` (or `install` if you want to merge the `cp` and `chmod`) it into the initrd?

rlaager commented on contrib/initramfs/hooks/zfsunlock.in:
```diff
@@ -0,0 +1,65 @@
+done
+EOF
+	chmod 0755 "$DESTDIR"/bin/zfsunlock
+	echo "zfsunlock" >> "$DESTDIR"/etc/profile
```
This (appending to `/etc/profile`) still feels wrong and I'd like to know if there are better options. I realize it's in the initrd, which makes it far less bad, of course, than if that were on the running system.

felixdoerre replied:
I think that any mechanism that "captures" all shells on the system is a problem. Dropbear, might not be used to do encryption unlocks but to do other things in the initramfs (e.g. remotely debugging a boot process). I think sticking to what the normal "cryptroot" does should be enough: Not forcing its way into the users foreground and if a user wants to have a unlock command directly triggered with the ssh-session they can just use ssh force-commands with their ssh-key in dropbear's authorized keys: e.g. prefix the key with `command="zfsunlock" ssh-rsa ....`.

rlaager replied:
Right, a forced command is a good way for the admin to make it automatic. Otherwise they can just run a command on login.

rlaager commented on contrib/initramfs/hooks/zfsunlock.in:
```diff
@@ -0,0 +1,65 @@
+		echo "password for $zfs_fs_name accepted"
+		zfs_console_askpwd_pid=$(ps a -o pid= -o args | grep -v grep | grep $zfs_console_askpwd_cmd | cut -d ' ' -f3 | sort -n | head -n1)
+                echo -e "resuming boot sequence\nplease reconnect in a while ..."
+		if [ ! -z $zfs_console_askpwd_pid ]; then
```
`if [ -n ... ]`

andrey42 commented:
Updated version, now the zfsunlock is removed from /etc/profiles, Makefile updated, zfsunlock script no longer processed by sed and copied as is via initrd Makefile. I've left initramfs hook script creating zfsunlock as it is, because as per feedback from users,  there were cases where dropbeap was absent from binary folder due to dropbear-initramfs dependencies broken in some Debian-based derivatives.

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+		entry_msg_shown=true
+	fi
+	sleep 1;
+	echo -n "."
```
Is the constant "." necessary? It's not a huge problem, but if the system sits for a while, that's going to fill the screen. But this is also what is requiring `echo -n` which is non-POSIX (albeit supported by even things like `dash`).

There are a number of `echo -n` instances that could probably be plain `echo`. I'm not seeing why these various files can't have a newline at the end.

andrey42 replied:
removed "."

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+. /usr/share/initramfs-tools/hook-functions
+
+if [ -r "${DESTDIR}/sbin/dropbear" ]; then
+	cat >"$DESTDIR"/bin/zfsunlock <<'EOF'
```
I'd still like to see this as a separate file. It'd be easier to work with. For example, I could run `shellcheck` on it trivially.

On that note, can you run shellcheck on your various scripts and address any non-bogus issues it finds?

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+
+if [ -r "${DESTDIR}/sbin/dropbear" ]; then
+	cat >"$DESTDIR"/bin/zfsunlock <<'EOF'
+#!/bin/sh
```
Newly written shell scripts should use `sh -eu` if at all possible. This will require you to initialize `$entry_msg_shown`.

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+	cat >"$DESTDIR"/bin/zfsunlock <<'EOF'
+#!/bin/sh
+while [ ! -e /run/zfs_fs_name ] && [ ! -e /run/zfs_unlock_complete ]; do
+	if [ ! $entry_msg_shown ]; then
```
Why is it necessary for the initial message to be inside the loop? Given the conditions of this loop and what is going on with those files, it's always going to fire. You could eliminate `entry_msg_show entirely` and just put the echo above the loop.

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+	sleep 1;
+	echo -n "."
+done
+if [ -e /run/zfs_unlock_complete ]; then
```
You could probably rewrite this whole thing to be just:
```
echo "Waiting for zfs init script to complete, press Ctrl-C to exit into shell..."
while : ; do
    if [ -e /run/zfs_unlock_complete ] ; then
        exit 0
    fi
    sleep 1
done

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+zfs_fs_name=""
+fs_password_accepted=false
+while true; do
+	while [ $fs_password_accepted == "true" ] && [ $(cat /run/zfs_fs_name) == "$zfs_fs_name" ] && [ ! -e /run/zfs_unlock_complete ]; do
```
`==` is not POSIX. Use `=`. Likewise below.

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+	zfs_fs_name=$(cat /run/zfs_fs_name)
+	zfs_console_askpwd_cmd=$(cat /run/zfs_console_askpwd_cmd)
+	systemd-ask-password "Encrypted ZFS password for ${zfs_fs_name}:" | \
+		/sbin/zfs load-key $zfs_fs_name
```
`$zfs_fs_name` needs quoting here and below. This is not even hypothetical; we have seen problems from scripts not quoting when they run into datasets having spaces.

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+	if [ $(/sbin/zfs get -H -ovalue keystatus $zfs_fs_name 2> /dev/null) == "available" ]; then
+		fs_password_accepted=true
+		echo "password for $zfs_fs_name accepted"
+		zfs_console_askpwd_pid=$(ps a -o pid= -o args | grep -v grep | grep $zfs_console_askpwd_cmd | cut -d ' ' -f3 | sort -n | head -n1)
```
Is busybox `ps` outputting something different than GNU `ps`? If I run `ps a -o pid= -o args | grep -v grep`, I get the PID in column 1, not 3 (as you're using in cut).

andrey42 replied:
yep, there is a difference, noted it during tests. Busybox ps version misses several output features and has diffirent columns output sequence, confirned it on Debian 10 and Ubuntu 18.4

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+		fs_password_accepted=true
+		echo "password for $zfs_fs_name accepted"
+		zfs_console_askpwd_pid=$(ps a -o pid= -o args | grep -v grep | grep $zfs_console_askpwd_cmd | cut -d ' ' -f3 | sort -n | head -n1)
+                echo -e "resuming boot sequence\nplease reconnect in a while ..."
```
`echo -e` is not POSIX. Use two `echo`s.

nit: Also, I'd modify the case and punctuation slightly:
```
echo "Resuming boot sequence."
echo "Please reconnect in a while."
```

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+		/sbin/zfs load-key $zfs_fs_name
+	if [ $(/sbin/zfs get -H -ovalue keystatus $zfs_fs_name 2> /dev/null) == "available" ]; then
+		fs_password_accepted=true
+		echo "password for $zfs_fs_name accepted"
```
nit: `"Password for $zfs_fs_name accepted."

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+	exit 0
+fi
+echo
+echo "Unlocking encrypted ZFS pools"
```
nit: I'd probably add a "..." here.

rlaager commented on contrib/initramfs/hooks/zfsunlock:
```diff
@@ -0,0 +1,64 @@
+fi
+echo
+echo "Unlocking encrypted ZFS pools"
+echo "Enter correct password or press Ctrl-C to exit into shell"
```
nit: Add a period to the end of this sentence. I'd also add "the" in front of "correct". And I might drop "correct" anyway.

rlaager commented on contrib/initramfs/scripts/zfs:
```diff
@@ -435,6 +439,7 @@ decrypt_fs()
 			# Prompt with ZFS tty, otherwise
 			else
 				# Setting "printk" temporarily to "7" will allow prompt even if kernel option "quiet"
```
nit: Reword to something like: Temporarily setting "printk" to "7" allows the prompt to appear even when the "quiet" kernel option has been used.

andrey42 commented:
Updated script to match requested changes. ps utility inside buzybox do differs from GNU version, found it durign the tests, hence the "special treatment". Also, as per request, I've moved echoing out of the cycle, but left the conditions check for /run/zfs_fs_name and /run/zfs_unlock_complete, because as per my tests, there is a race condition between these two flags on low-performing system: the script might be called at the point in time where: none of the flags is present yet, either only /run/zfs_fs_name can appear, both /run/zfs_fs_name and /run/zfs_unlock_complete can appear or /run/zfs_unlock_complete can appear. To cover the case for non-encrypted system, the script must transparently exit after waiting, when /run/zfs_unlock_complete has been created but no /run/zfs_fs_name flag is present.

rlaager commented on contrib/initramfs/zfsunlock:
```diff
@@ -0,0 +1,41 @@
+set -eu
+echo "Waiting for zfs init script to complete, press Ctrl-C to exit into shell ..."
+while [ ! -e /run/zfs_fs_name ] && [ ! -e /run/zfs_unlock_complete ]; do
+	sleep 1;
```
no semicolon needed here or in the sleep below

rlaager replied:
I understand more what you're doing now. You're waiting as long as _neither_ of these exist. But you could still do this as:

```
while [ ! -e /run/zfs_fs_name ]; do
    if [ -e /run/zfs_unlock_complete ]; then
	exit 0
    fi
    sleep 1
done
```
which would have the same effects as now: If `/run/zfs_unlock_complete` shows up, you exit. If `/run/zfs_fs_name` shows up but not `/run/zfs_unlock_complete`, you break out of the loop.

If the filesystem is unencrypted, then `/run/zfs_fs_name` will never be created. Now that `zfsunlock` is run manually, that likely doesn't matter. But even if someone does run it, the check for `/run/zfs_unlock_complete` is sufficient.

rlaager commented on contrib/initramfs/zfsunlock:
```diff
@@ -0,0 +1,41 @@
+#!/bin/sh
+
+set -eu
```
Personally, I'd put this in the shebang: #!/bin/sh -eu

If you prefer it here, that works too.

rlaager commented on contrib/initramfs/zfsunlock:
```diff
@@ -0,0 +1,41 @@
+	zfs_fs_name=$(cat /run/zfs_fs_name)
+	zfs_console_askpwd_cmd=$(cat /run/zfs_console_askpwd_cmd)
+	systemd-ask-password "Encrypted ZFS password for ${zfs_fs_name}:" | \
+		/sbin/zfs load-key "$zfs_fs_name"
```
I haven't tested `zfs load-key`'s exit status. If you enter the wrong password, does it exit non-zero? If so, the new `set -e` is going to break this in the case of an incorrect password.

andrey42 replied:
yep, it does, there are 3 diffirent exit codes, one for incorrect key, one for master key unavailable and one for uncaught exception

rlaager commented on contrib/initramfs/zfsunlock:
```diff
@@ -0,0 +1,41 @@
+zfs_fs_name=""
+fs_password_accepted=false
+while true; do
+	while [ "$fs_password_accepted" = "true" ] && [ "$(cat /run/zfs_fs_name)" = "$zfs_fs_name" ] && [ ! -e /run/zfs_unlock_complete ]; do
```
Can you elaborate a bit on what's going on here or confirm if my understanding is correct? Is this saying: if the user's password was accepted, sleep until either the unlocking is done or until `/run/zfs_fs_name` changes (i.e. we have another filesystem to unlock)?

If so, this would probably be cleaner to put in the code path where the password is accepted. That would eliminate the need to check `$fs_password_accepted` in this loop. Plus, since that's the only use for that variable, it could be removed entirely. So it would be something like.

```
...
echo "Please reconnect in a while."
... the kill thing ...
# Wait for another filesystem to unlock.
while [ "$(cat /run/zfs_fs_name)" = "$zfs_fs_name" ]; do
    if [ -e /run/zfs_unlock_complete ]; then
        break # or exit 0, your choice
    fi
    sleep 1
done
```

Then, the first check in the main loop is for `/run/zfs_unlock_complete` so your `while true` can become `while [ ! -e /run/zfs_unlock_complete]`.

rlaager replied:
Maybe I forgot to mark this unresolved.

andrey42 replied:
I've already moved the messages output and break to anotehr place.

rlaager commented on contrib/initramfs/zfsunlock:
```diff
@@ -0,0 +1,41 @@
+	exit 0
+fi
+echo
+echo "Unlocking encrypted ZFS pools ..."
```
nit: Instead of "pools", "filesystems".

rlaager commented on contrib/initramfs/zfsunlock:
```diff
@@ -0,0 +1,41 @@
+		zfs_console_askpwd_pid=$(ps a -o pid= -o args | grep -v grep | grep "$zfs_console_askpwd_cmd" | cut -d ' ' -f3 | sort -n | head -n1)
+                echo "Resuming boot sequence."
+		echo "Please reconnect in a while."
+		if [ -n "$zfs_console_askpwd_pid" ]; then
```
For clarity, I would put the lookup of the PID and the killing of it next to each other, and the echos next to each other, rather than mixing them.

rlaager commented on contrib/initramfs/hooks/zfsunlock.in:
```diff
@@ -0,0 +1,20 @@
+
+. /usr/share/initramfs-tools/hook-functions
+
+if [ -r "${DESTDIR}@sbindir@/dropbear" ]; then
```
Why is this check required? You have listed dropbear as a prereq.

rlaager commented on contrib/initramfs/scripts/zfs:
```diff
@@ -405,6 +405,8 @@ decrypt_fs()
 		ENCRYPTIONROOT="$(get_fs_value "${fs}" encryptionroot)"
 		KEYLOCATION="$(get_fs_value "${ENCRYPTIONROOT}" keylocation)"
 
+		echo -n "${ENCRYPTIONROOT}" > /run/zfs_fs_name
```
I'm not sure that you ever answered whether the `-n` is really required on these. From my testing, they are not:

```
echo foo > /run/zfs_fs_name
zfs_fs_name=$(cat /run/zfs_fs_name)
echo x${zfs_fs_name}y
```

With `dash` and `bash`, I get "xfooy" with no space or newline.

andrey42 commented:
After tests, I've removed -e option, zfs load-key does return non-zero exit code on wrokg password, breaking the script. Moved the message about reconnecting into out of break condition, it it requored for case where are more than one system to unlock and there is a racing condition due to sluggish system

andrey42 commented:
Dear @behlendorf  Could you please explain why buiddbot builds are failing for my commit ?
Should I rebase commit once more to latest master branch ?
I'm asking because on my Debian package build & tests workflow using Ubuntu 18.04 KVM image based on Gitbug actions and commit pushed as a patch I've got no errors during the build & tests using latest master branch.
See more details here: https://github.com/andrey42/zfs-debian/actions/runs/89435559

tcaputi commented:
@andrey42 i believe you just need to add the `zfs/contrib/initramfs/zfsunlock` script to `zfs/contrib/initramfs/Makefile.am` to ensure that it gets added to the built packages.

rlaager commented on contrib/initramfs/scripts/zfs:
```diff
@@ -405,6 +405,8 @@ decrypt_fs()
 		ENCRYPTIONROOT="$(get_fs_value "${fs}" encryptionroot)"
 		KEYLOCATION="$(get_fs_value "${ENCRYPTIONROOT}" keylocation)"
 
+		echo -n "${ENCRYPTIONROOT}" > /run/zfs_fs_name
```
Still an `echo -n` here.

andrey42 replied:
Removed, works correctly after tests, thans for reminding

rlaager commented on contrib/initramfs/zfsunlock:
```diff
@@ -0,0 +1,43 @@
+#!/bin/sh
+
+set -u
```
`set -e` has disappeared.

andrey42 replied:
put || true next to load-key, how non-null exit code does not cause script to exit, so put -e back

Sithuk commented:
> 
> 
> After tests, I've removed -e option, zfs load-key does return non-zero exit code on wrong password, breaking the script. Moved the message about "unlock complete" into break condition, it is required for case where are more than one system to unlock and there is a racing condition due to sluggish system response. To prevent this, added sync over pipe to prevent race condition, without it there were numerous cases when "unlock complete" message was not shown due to premature disconnection.

Would appending "|| true" to the end of the zfs load-key command allow you to re-instate "set -e"?

andrey42 commented:
Removed extra -n for filesystem. I've removed -e bash option, because with it, zfsunlock script immediately crashes due to non-zero exit code returned by zfs load-key.

gmelikov commented:
@andrey42 you may want to remain `set -e` and catch expected errors instead of silently ignore any error, it's safer, for ex.
```
set -e
EXIT_CODE=0
command || EXIT_CODE=$?
echo $EXIT_CODE
```

andrey42 commented:
> 
> 
> @andrey42 you may want to remain `set -e` and catch expected errors instead of silently ignore any error, it's safer, for ex.
> 
> ```
> set -e
> EXIT_CODE=0
> command || EXIT_CODE=$?
> echo $EXIT_CODE
> ```

Yep, adding || true worked too, tested on Debian 10 & Ubuntu 18.4
added back -e option

andrey42 commented on contrib/initramfs/zfsunlock:
```diff
@@ -0,0 +1,43 @@
+	if [ -e /run/zfs_unlock_complete ]; then
+		echo "Unlocking complete, resuming boot sequence."
+		echo "Please reconnect in a while."
+		echo "ok" > /run/zfs_unlock_complete_notify
```
There is a racing condition once console input is killed, and several tests confirmed the dropbear disconnects before text about being disconnection is displayed to the user. This creates false assumption about the session crash. I have added pipeline wait/notify to prevent this.

rlaager replied:
That makes sense. Is there a particular reason some of this uses the creation/existence of a file as the flag and this thing uses a FIFO? Can they all be done in the same way? If not, that's fine.

andrey42 replied:
wait/notify is sure better than checking flags, but the problem here is user hitting Ctrl-C during the password input at random intervals. After several use cases I've removed pipelines for syncronization during the password input, because I've discovered hitting Ctrl-C sometimes led to  pipeline stuck waiting for input. So, I chose to use wait/notify only when all interactions are done, i.e. on final stage.

andrey42 commented on contrib/initramfs/Makefile.am:
```diff
@@ -21,3 +22,6 @@ install-initrdSCRIPTS: $(EXTRA_DIST)
 		cp $(top_builddir)/contrib/initramfs/$$d/zfs \
 		    $(DESTDIR)$(initrddir)/$$d/; \
 	done
+	$(MKDIR_P) $(DESTDIR)$(initrddir)/hooks; \
```
Yep, looks obsolete, will remove it.

rlaager commented on contrib/initramfs/Makefile.am:
```diff
@@ -21,3 +22,5 @@ install-initrdSCRIPTS: $(EXTRA_DIST)
 		    $(DESTDIR)$(initrddir)/$$d/; \
 	done
+	cp $(top_builddir)/contrib/initramfs/zfsunlock \
+	    $(DESTDIR)/$(initrddir)/;
```
This semi-colon is not necessary.

andrey42 replied:
removed

rlaager commented:
We seem to be talking past enough other on that one point. I've added a (untested) commit which shows the change I'm proposing. If this looks good to you, you can squash it in.

andrey42 commented:
> We seem to be talking past enough other on that one point. I've added a (untested) commit which shows the change I'm proposing. If this looks good to you, you can squash it in.
Squashed and pushed new version, added conditional check to skip existing pipe, happens if user press Ctrl-C during password entry, then runs the script again.

sergeymakinen commented:
I'm not sure if it's worth noting but: tested on Proxmox 6.1 (where this PR is very welcome) with systemd-boot (EFI) and its initramfs lacks `systemd-ask-password` so I had to add `copy_exec /usr/bin/systemd-ask-password /usr/bin/` to make `zfsunlock` work.

Fabian-Gruenbichler commented:
> I'm not sure if it's worth noting but: tested on Proxmox 6.1 (where this PR is very welcome) with systemd-boot (EFI) and its initramfs lacks `systemd-ask-password` so I had to add `copy_exec /usr/bin/systemd-ask-password /usr/bin/` to make `zfsunlock` work.

it's in /bin/systemd-ask-password (possibly symlinked to /usr/bin), and the [COPY_EXEC_LIST](https://github.com/openzfs/zfs/pull/10027/files#diff-006e0b3f7232bcda0e15903882992566R24) should take care of copying it into the initrd? you can verify with `lsinitramfs`.

Debian (and therefor Proxmox) don't run systemd in their initrd though, but a custom shell init, so not sure whether systemd-ask-password even works (well).

sergeymakinen commented:
> it's in /bin/systemd-ask-password (possibly symlinked to /usr/bin), and the [COPY_EXEC_LIST](https://github.com/openzfs/zfs/pull/10027/files#diff-006e0b3f7232bcda0e15903882992566R24) should take care of copying it into the initrd? you can verify with `lsinitramfs`.

Ah, got it, thanks! I tested the PR manually and that's why I didn't have `systemd-ask-password`. :)

> Debian (and therefor Proxmox) don't run systemd in their initrd though, but a custom shell init, so not sure whether systemd-ask-password even works (well).

After it's copied, it works. At least in my tests.

andrey42 commented:
> 
> 
> I'm not sure if it's worth noting but: tested on Proxmox 6.1 (where this PR is very welcome) with systemd-boot (EFI) and its initramfs lacks `systemd-ask-password` so I had to add `copy_exec /usr/bin/systemd-ask-password /usr/bin/` to make `zfsunlock` work.

There is a zfs hook for initrams located here: contrib/initramfs/hooks/zfs.in
it copies systemd-ask-password from the location retrieved via PATH variable

````
COPY_EXEC_LIST="$COPY_EXEC_LIST $(which systemd-ask-password)"
````

So, add systemd-ask-password to PATH env variable, (should be there anyway) and it will be copied into initramfs image.

andrey42 commented:
> 
> 
> I tested this on a buster system. In the course of doing that, I ended up down some big rabbit holes. I've pushed some changes that are intertwined with this. Can you take a look, at least at the commits starting with yours? If you're happy, please squash the last commits into yours, but leave the two commits before it as separate commits. If this still passes your testing, you're happy with it, and you've squashed those, then I'm ready to give it my approval.

Looks good to me, these are mainly cosmetical changes, no impact on algorith itself.
Pushed merged commit.

felixdoerre commented:
Just as a note: I noticed that in debian unstable, in the `dropbear-initramfs`-package there is this check which is required to have dropbear installed into the initramfs:
```
[ -r /etc/crypttab ] || exit 0
```
so you probably need `cryptroot` or something similar installed and at least an empty `crypttab` for dropbear to be installed. I'd consider this a bug in `dropbear-initramfs` which should be followed up upon. Do you all have empty `crypttab`-files or is the check missing in the initramfs-hook on your test installations?

rlaager commented:
> so you probably need `cryptroot` or something similar installed and at least an empty `crypttab` for dropbear to be installed. I'd consider this a bug in `dropbear-initramfs` which should be followed up upon. Do you all have empty `crypttab`-files or is the check missing in the initramfs-hook on your test installations?

Good catch! I have a stub `crypttab` file (not one that I created).

> I checked and the annoying check is removed in Debian buster, so this is not a problem in newer Debian versions.

That's probably fine then. This code is only going to show up in newer releases anyway. If it gets backported somewhere (or the user builds from source), it's easy enough to fix.

tcaputi commented:
> That's probably fine then. This code is only going to show up in newer releases anyway. If it gets backported somewhere (or the user builds from source), it's easy enough to fix.

So if I wanted to install patch this onto my Ubuntu 20.04 machine (building from source), what would I need to do other than just `make pkg`?

andrey42 commented:
> > That's probably fine then. This code is only going to show up in newer releases anyway. If it gets backported somewhere (or the user builds from source), it's easy enough to fix.
> 
> So if I wanted to install patch this onto my Ubuntu 20.04 machine (building from source), what would I need to do other than just `make pkg`?

You can use my Debian repo, it does contain ZFS Debian packages with ZSTD patches and unlock ZFS patches installed.  I've used [Debian repo](https://salsa.debian.org/zfsonlinux-team/zfs) as a base for packaging topology, so my packages can be installed on any Debian derivatives. 

Also, if you clone the repo and add your own signing keys, then you can add/remove any patches you wish on your own, there is a Github Action workflow, that pulls openzfs/zfs master branch, applies patches, then builds, tests and creates complete set of ZFS Debian packages saved as Gihhub pages content, so there is no need for any PPA or additional testing infrastructure whatsoever.

Sithuk commented:
> 
> 
> > > That's probably fine then. This code is only going to show up in newer releases anyway. If it gets backported somewhere (or the user builds from source), it's easy enough to fix.
> > 
> > 
> > So if I wanted to install patch this onto my Ubuntu 20.04 machine (building from source), what would I need to do other than just `make pkg`?
> 
> You can use [my Debian repo](https://andrey42.github.io/zfs-debian/), it does contain ZFS Debian packages with ZSTD patches and unlock ZFS patches installed. I've used [original Debian team repository](https://salsa.debian.org/zfsonlinux-team/zfs) as a base for packaging topology, so my packages can be installed on any Debian derivatives.
> 
> Also, if you clone the repo (follow readme instructions) and add your own signing keys, then you can add/remove any patches you wish on your own, there is a Github Action workflow, that pulls openzfs/zfs master branch, applies patches, then builds, tests and creates complete set of ZFS Debian packages saved as Gihhub pages content, so there is no need for any PPA or additional testing infrastructure whatsoever.

Hi @andrey42 are you certain your repo contains this patch #10027? I installed but entering "zfsunlock" did nothing when I logged into dropbear. I couldn't find "zfsunlock" on the system either.

tcaputi commented:
@andrey42 I just tried installing this on my Ubuntu 20.04 personal server. I got this error:
```
root@roy-ubuntu:~# ssh root@192.168.1.100


BusyBox v1.30.1 (Ubuntu 1:1.30.1-4ubuntu6) built-in shell (ash)
Enter 'help' for a list of built-in commands.

# zfsunlock

Unlocking encrypted ZFS filesystems...
Enter the password or press Ctrl-C to exit.

Encrypted ZFS password for rpool: ************
Password for rpool accepted.
ps: invalid option -- 'a'
/usr/bin/zfsunlock: line 28: head: not found
BusyBox v1.30.1 (Ubuntu 1:1.30.1-4ubuntu6) multi-call binary.

Usage: ps

Show list of processes

        w       Wide output
        l       Long output
#
```

The error comes from this line in `zfsunlock`:
```
zfs_console_askpwd_pid=$(ps a -o pid= -o args | grep -v grep | grep "$zfs_console_askpwd_cmd" | cut -d ' ' -f3 | sort -n | head -n1)
```
It appears that busybox's ps does not have take an `a` parameter and `head` doesn't exist in busybox at all. This could probably be resolved by installing an environment other than busybox into the initramfs, but it is the default and the directions in the README don't indicate that I need to change anything.

Any thoughts about a good way to fix this?

rlaager commented:
`head` should be able to be fixed with another `copy_exec`.

rlaager commented:
So this is interesting. On Ubuntu, if you have the `busybox-static` package installed, you get a busybox that supports `ps a` and `head`. If you only have `busybox-initramfs`, you don't. This is reproducible on a running system like this:

```
/usr/lib/initramfs-tools/bin/busybox ps a -o pid= -o args
vs
/bin/busybox ps a -o pid= -o args
```

Debian, at least on Buster, does not have a `busybox-initramfs`. It has a `busybox` package that supports both `ps a` and `head`.

@tcaputi / @andrey42 Does this work: `ps | grep -v grep | grep "$zfs_console_askpwd_cmd" | cut -d' ' -f1 | sort -n | tr "\n" " " | cut -d" " -f1`

tcaputi commented:
Testing right now....

tcaputi commented:
@rlaager It looks like cut isn't working the way you'd like it...

```
# ps | grep -v grep | grep "load-key" | cut -d' ' -f1

# ps | grep -v grep | grep "load-key"
  825 root      8952 S    /sbin/zfs load-key rpool
```

tcaputi commented:
awk exists (at least in my busybox). So this seems to work:
```
# ps | grep -v grep | grep "load-key" | awk '{ print $1 }'
825
# ps | grep -v grep | grep "load-key" | awk '{ print $1 }' | sort -n | tr "\n" " " | cut -f1
825
#
```

rlaager commented:
If you have awk, then this should work:
`ps | grep -v grep | awk "/$zfs_console_askpwd_cmd/ { print \$1 }" | sort -n | awk '{print}'`
or this:
`ps | awk '!'"/grep/ && /$zfs_console_askpwd_cmd/ { print \$1 }" | sort -n | awk '{print}'`

tcaputi commented:
I'm not sure what the point of the last `awk` is in either example, but I'm guessing thats just because I'm not very good at awk :P

rlaager commented:
It was supposed to replace the `head`. But it needs to be `awk '{print;exit}'`. That said, busybox `ps` seems to be outputting the pids sorted anyway. I'm not sure if we can rely on that. If we can, `ps | awk '!'"/grep/ && /$zfs_console_askpwd_cmd/ { print \$1; exit }"` is fine.

tcaputi commented:
I'm not sure if that will work with normal `ps`...
```
root@roy-ubuntu:~# ps aux | awk '!'"/grep/ && /sleep/ { print \$1; exit }"
root
```

We might need to try to detect what kind of ps we're running.

rlaager commented:
It's not meant to work with normal `ps`. This is specifically for the initramfs environment, right? Does that ever end up with GNU `ps`?

rlaager commented:
```
$ sleep 30 &
[2] 14096
$ sleep 30 &
[3] 14157
$ zfs_console_askpwd_cmd=sleep ; /usr/lib/initramfs-tools/bin/busybox ps | /usr/lib/initramfs-tools/bin/busybox awk '!'"/grep/ && /$zfs_console_askpwd_cmd/ { print \$1; exit }"
14096
$ zfs_console_askpwd_cmd=sleep ; /bin/busybox ps | /bin/busybox awk '!'"/grep/ && /$zfs_console_askpwd_cmd/ { print \$1; exit }"
14096
```

tcaputi commented:
@rlaager Right. I'm just saying that if anyone (for whatever reason) has a real version of `ps` in their initramfs it won't work. Probably a very small percentage of users though.

andrey42 commented:
> 
> 
> @rlaager Right. I'm just saying that if anyone (for whatever reason) has a real version of `ps` in their initramfs it won't work. Probably a very small percentage of users though.

Thanks for noting it. Better to cover all possible angles.
 I was withessing many cases where dropbear and busybox was installed manually via dpkg, breaking dependencies, when people use low-quality instructions floating around on installing dropbear manually.

innerand commented:
Hi, 

as the documentation is still todo, could somebody give me a hint on how to use this with Debian Buster? 
I mean is this a package, is it already available in repositories or should the files be copied manually (which?)? 
Unfortunately most links from the discussions above are dead. 

Thanks a lot! 
innerand

terem42 commented:
The code has been merged in master repo, so you can cherrypick then add patch from this PR to the source code, then compile the ZFS package, for example latest version 0.8.5. The actual code will be part of 2.0 package or earlier as per users request.
To use it, all you need is to install dropbear-initramfs package and add your authorized keys to the dropbear folder. 

````
 cp /home/myuser/.ssh/authorized_keys /etc/dropbear-initramfs/authorized_keys
````

Then on every encrypted system boot, the system will open SSH connection and wait for keys to entered via the SSH by human user or via script. At the same time, if you enter ZFS password into console, SSH server is quitetly shut down, and boot resumes in usual manner, i.e. entering ZFS encryption password via SSH and console can safely coexists.

 This solution is extremely usable in cases where you cannot trust credentials storage of your datacenter vendor (AWS, Azure, Google etc.) and need to rotate ZFS encryption keys without any hosting vendor involvement along automation of  systems boots, automatically excluding machines from the distributed pool in case the system is compromised, or there is a threat it might be.

Fmstrat commented:
@terem42  I'm not sure how you'd like to receive user requests, but I would very much like to see this in a current version vs waiting for 2.0. Not as much as raidz expansion, but it's up there on my list to replace my current mdadm workflow.

terem42 commented:
Well, I'm not using official distributions anyway. Normally I take master and apply my own set of patches and PRs.
If you use Ubuntu 18.4 or Debian 10 for Prod systems, you can fork my repo https://github.com/terem42/zfs-debian
or use it as a unofficial Debian repo, it does include this commit as well as others. The workflow incluses full-blown set tests on each commit, so preliminary zfs packaje stability is assured.

n0rc commented:
ZFS 2.0 is also already available in Debian 10 backports.
